File: ros2_ws/src/py_msg/py_msg/robot_status_subscriber.py
from typing import Optional

import rclpy
from rclpy.node import Node
from tutorial_interfaces.msg import RobotStatus


class RobotStatusSubscriber(Node):
    def __init__(self) -> None:
        super().__init__("robot_status_subscriber")
        self.subscription = self.create_subscription(
            RobotStatus, "robot_status", self.listener_callback, 10
        )

    def listener_callback(self, msg: RobotStatus) -> None:
        self.get_logger().info(
            f"Received from {msg.robot_name}: "
            f"battery={msg.battery_level:.1f}, "
            f"joints={msg.joint_positions}"
        )


def main(args: Optional[list[str]] = None) -> None:
    rclpy.init(args=args)
    node = RobotStatusSubscriber()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_ws/src/cpp_msg/src/robot_status_subscriber.cpp
#include "cpp_msg/robot_status_subscriber.hpp"

#include <sstream>

namespace learning_ros2::cpp_msg {

RobotStatusSubscriber::RobotStatusSubscriber(const rclcpp::NodeOptions& options)
    : rclcpp::Node("robot_status_subscriber", options) {
  subscription_ = this->create_subscription<RobotStatus>(
      "robot_status", 10,
      std::bind(&RobotStatusSubscriber::topic_callback, this,
                std::placeholders::_1));
}

void RobotStatusSubscriber::topic_callback(
    const RobotStatus::SharedPtr msg) const {
  std::stringstream joints_ss;
  joints_ss << "[";
  for (size_t i = 0; i < msg->joint_positions.size(); ++i) {
    joints_ss << msg->joint_positions[i];
    if (i + 1 < msg->joint_positions.size()) {
      joints_ss << ", ";
    }
  }
  joints_ss << "]";

  RCLCPP_INFO(this->get_logger(),
              "Received from %s: battery=%.1f, active=%s, joints=%s",
              msg->robot_name.c_str(), msg->battery_level,
              msg->is_active ? "true" : "false", joints_ss.str().c_str());
}

}  // namespace learning_ros2::cpp_msg
